lib/pbench/test/unit/server/test_datasets_metadata.py

```diff
@@ -0,0 +1,178 @@
+from http import HTTPStatus
+import itertools
+
+from pbench.server.api.resources import ParamType
+from pbench.server.api.resources.datasets_metadata import DatasetsMetadata
+
+
+class TestDatasetsMetadata:
+    def test_get_no_dataset(self, client, server_config, attach_dataset):
+        response = client.get(
+            f"{server_config.rest_uri}/datasets/metadata",
+            query_string={
+                "controller": "node",
+                "name": "foobar",
+                "metadata": ["seen", "saved"],
+            },
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json == {"message": "Dataset foobar not found"}
+
+    def test_get_bad_keys(self, client, server_config, attach_dataset):
+        response = client.get(
+            f"{server_config.rest_uri}/datasets/metadata",
+            query_string={
+                "controller": "node",
+                "name": "foobar",
+                "metadata": ["xyzzy", "plugh", "owner", "access"],
+            },
+        )
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json == {"message": "Dataset foobar not found"}
+
+    def test_get(self, client, server_config, provide_metadata):
+        response = client.get(
+            f"{server_config.rest_uri}/datasets/metadata",
+            query_string={
+                "controller": "node",
+                "name": "drb",
+                "metadata": ["seen", "saved", "access", "user"],
+            },
+        )
+        assert response.status_code == HTTPStatus.OK
+        assert response.json == {
+            "saved": False,
+            "seen": True,
+            "access": "private",
+            "user": {"contact": "me@example.com"},
+        }
+
+    def test_put_missing_json_object(self, client, server_config, pbench_token):
+        """
+        Test behavior when no JSON payload is given
+        """
+        response = client.put(f"{server_config.rest_uri}/datasets/metadata")
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json.get("message") == "Invalid request payload"
+
+    def test_put_missing_keys(self, client, server_config, user_ok, pbench_token):
+        """
+        Test behavior when JSON payload does not contain all required keys.
+
+        Note that Pbench will silently ignore any additional keys that are
+        specified but not required.
+
+        TODO: This is mostly copied from commons.py
+        """
+        classobject = DatasetsMetadata(client.config, client.logger)
+
+        def missing_key_helper(keys):
+            response = client.post(
+                f"{server_config.rest_uri}/datasets/metadata", json=keys,
+            )
+            assert response.status_code == HTTPStatus.BAD_REQUEST
+            missing = [k for k in required_keys if k not in keys]
```

Comment:
Here's another nice spot for `set` manipulation:
```
            missing = list(set(required_keys) - set(keys))
```
(and, you can leave off the `list()`, because `str.join()` will happily accept a `set` in lieu of a `list`).